perceval/simulators/_simulator_utils.py

```diff
@@ -85,3 +88,20 @@ def _unitary_components_to_circuit(component_list: list, m: int = 0):
     for r, c in component_list:
         circuit.add(r, c)
     return circuit
+
+
+def _split_by_photon_count(sv: StateVector) -> SVDistribution:
```

Comment:
It can be very costly to create 1 SVD per SV in a noisy input state. Indeed, currently the `Source` generates a huge SVD (around 4**n elements) containing SV being only one annotated `FockState` at a time. When iterating on it, you'd create everytime one SVD containing exactly 1 element which is going to be hashed in order to be inserted in the dictionary (the SVD being a dict).

One hash call is fast, but 4**n hash calls, can hinder performances a lot.

I see in *simulator.py* that `_split_by_photon_count` is called everytime so:
Did you benchmark your changes on a regular noisy simulation with 8-10 photons? Is it faster than the current algorithm?

Reply:
After benchmarking, this version was indeed way worse.
For 9 photons with an imperfect source (brightness = 0.8, indistinguishability = 0.9, g2 = 0.1), for the whole `probs` call (not taking into account the SVD generation):
- From previous version: 8.08 s
- From the reviewed version: 24.3 s
- After upgrading (commit `a4ba7db`): 7.27 s

The new version is now slightly better than the previous one even with no superposed states